### ISSUE TITLE
Classic TODO example with events

### ISSUE THREAD

**Issue by bedeho** [mobxjs/mobx#1298]
## Premise
Consider the classic TODO example, with a `TodoStore` and `TodoItems`.
Imagine its part of some large multiscene application, where the user
is not always viewing the item list. Lastly assume that there will
at some point be some spontaneous event on each item which alters its state.

## Problem
A normal UX idiom is to show some sort of alert counter associated with the
list when a user is on some other scene. It will track the total number of times
such state changes have occured while the user is not looking at the list,
and reset the counter when the user navigates to the list.

How would one implement this in MOBX without resorting to trying to use normal
events and managed subscription to such events on stores?

## Proposed solution

We introduce an observable tracking the total number of such events through
the full lifetime of the `TodoStore`. Lets call it `totalEventCount`, and it would
be reactively maintained by doing something like

```js
// In `TodoStore`

@observable totalEventCount

addTodoItem(todoItem) {
  this.todos.push(todoItem)

  when('',
    () => {
      todoItem.state === 'this-is-my-funny-state'
    },
    () => {
      this.totalEventCount = this.totalEventCount + 1
    }
  )
}
```

In the user interface store, lets call it `UIStore`, which manages what scene is rendered in the app
we keep track of two observables

 - `this._totalCountWhileViewingListLastTime` manged by

```js
// In `UIStore`
constructor(todoStore) {

  this._todoStore = todoStore
  this._totalCountWhileViewingListLastTime = observable(0)

  autorun(() => {
    if(this.activeScene === UIStore.SCENE.TODOLIST) // Check if we are on todo list scene
      this._totalCountWhileViewingListLastTime = this._todoStore.totalEventCount
  })
}
```

- `numberEventsInTheBackground` computed by

```js
// In `UIStore`
@computed numberEventsInTheBackground() {
    return this._todoStore.totalEventCount - this._totalCountWhileViewingListLastTime
}
```

## Question

1.
Is this the right approach for solving this problem? While I am a novice, I found it extremely hard to arrive at this solution, so I am concerned that I may be overcomplicating
this some how.

2.
If this is on the right track, is there some documentation which perhaps conveys some of the underlying principles which would more easily allow someone to arrive at this sort of approach?

**Comment by urugator** [comment]
I would probably go with `EvenCounter` for each scene, pass it as dependency to relevant to stores (`TodoStore`, `TodoItem`, ...) and reset the counter assosciated with scene in action which changes the scene, so roughly:
```javascript
// in TodoStore
@action addTodoItem(todoItem) {
  this.todos.push(new TodoItem(this.eventCounter));
  this.eventCounter.incremenet();
}
// in TodoItem
@action setFunnyState(funnyState) {
  this.funnyState = funnyState;
  this.eventCounter.incremenet();
}
// In UIStore
@action changeScene(name) {
  this.scene = this.getSceneByName(name);
  this.scene.eventCounter.reset();
}
```
If you insist on using events (I don't recommend) I suggest to resort to "normal events and managed subscription", because MobX reactions `when/reaction/autorun` behaves quite differently (they're synchronous, but not immediate and they can't run as a part of a single TX)

**Comment by bedeho** [comment]
Thank you to both of you. It would certainly be nice to be able to factor out some reusable component that  could a work across multiple scenes like that.

**Comment by bedeho** [comment]
It seems like mutating state inside a computed value, as you are doing for `changeCount`, is discouraged, and perhaps not possible in strict mode, see second paragraph mobxjs/mobx-react#383 (comment).

`
Don't use computed to update state. It violates one way data flow (hard to follow, prone to errors), it may throw in strict mode, it may cause unwanted subscriptions (completedTodosCount subscribed to status.changed), it may cause staleness when computed operates in "non-reactive mode" (when not observed)...
`

**Comment by bedeho** [comment]
## Improvement

One issue with the approach I outlined, is that we had to introduce `totalEventCount` in the main `TodoStore`. This means that, as we make our UI richer, we may need to add more and more of these cumalative variables and reactive `when` handlers inside the `TodoStore`. This seems like it violates the separation of concerns, and is not so easy to manage.

What if we instead just have a single new observable array inside the `TodoStore`, called `itemsAdded`, which is just a list of all items ever added, even including ones that were eventually removed. This single observable can be observed by any part of the UI store hierarchy, and at the right spot one may introduce the same sort of `when` deduction as we currently have put in `TodoStore.addTodoItem`. We keep this local reaction local, and also never need to touch the `TodoStore` again, regardless of how we change the UI or the `TodoItem`.

Does this sound reasonable?

**Comment by spion** [comment]
`when` has very similar problems to mutating in a `computed` in that it makes it easier to create cycles.

The alternative that urugator suggested doesn't have the problem - instead of monitoring the active tab it has a reset method that gets called when the active tab changes.

We're not necessarily writing data models, but UI view-models. If a tab has multiple items reporting events, the change counter could be passed to multiple stores. If there are multiple counters, they would each need to keep a count anyway.

Another less performant, but also less coupled method would be to set an item's `item.updatedAt = Date.now()`. In that case the changeCounter would be

```
class ChangeCounter() {
  @observable dateReset = Date.now(), focused = false
  @action focus() { this.focused = true; }
  @action blur() { this.focused = false; this.dateReset = Date.now(); }
  constructor(countItems) { this.countItems = countItems }
  @computed get changeCount() { 
    return this.focused ? 0 : this.countItems(this.dateReset)
  }
}

let todoStore = new TodoStore()

let counter = new ChangeCounter(resetAt => todoStore.items.filter(item => item.updatedAt > resetAt).length)
```
This way the todo store itself doesn't know anything about change counting and doesn't use events.

**Comment by mweststrate** [comment]
Closing issue as answers have been provided and the issue is inactive now.
